File: examples/projects/benchmarks/dyn_dispatch/dyn_dispatch.cpp
// 基准: dyn_dispatch —— 2000 万次虚函数多态分派
// 与 dyn_dispatch.zeta 逻辑严格一致。输出 = 70000000
#include <cstdio>
#include <cstdint>

struct Shape {
    virtual int64_t sides() const = 0;
    virtual ~Shape() {}
};
struct Tri : Shape {
    int64_t n;
    explicit Tri(int64_t v) : n(v) {}
    int64_t sides() const override { return n + 3; }
};
struct Quad : Shape {
    int64_t n;
    explicit Quad(int64_t v) : n(v) {}
    int64_t sides() const override { return n + 4; }
};

int main() {
    Tri t{0};
    Quad q{0};
    Shape *d1 = &t;
    Shape *d2 = &q;
    int64_t sum = 0;
    for (int64_t i = 0; i < 10000000; i++) {
        sum += d1->sides();
        sum += d2->sides();
    }
    printf("%lld\n", sum);
    return 0;
}

File: examples/projects/benchmarks/dyn_dispatch/dyn_dispatch.swift
// 基准: dyn_dispatch —— 2000 万次多态分派（protocol 存在类型 witness table）
// 与 dyn_dispatch.rl 逻辑严格一致。输出 = 70000000
protocol Shape {
    func sides() -> Int64
}

struct Tri: Shape {
    let n: Int64
    func sides() -> Int64 { n + 3 }
}

struct Quad: Shape {
    let n: Int64
    func sides() -> Int64 { n + 4 }
}

let d1: any Shape = Tri(n: 0)
let d2: any Shape = Quad(n: 0)
var sum: Int64 = 0
for _ in 0..<10_000_000 {
    sum += d1.sides()
    sum += d2.sides()
}
print(sum)
